pallets/msa/src/lib.rs

```diff
@@ -1134,6 +1161,104 @@ pub mod pallet {
 			Self::deposit_event(Event::RecoveryProviderRemoved { provider_id: provider });
 			Ok(())
 		}
+
+		/// Recover an MSA account with a new control key.
+		/// This extrinsic is called by approved Recovery Providers after verifying
+		/// the user's Recovery Intermediary Hashes and Authentication Contact.
+		///
+		/// # Events
+		/// * [`Event::AccountRecovered`]
+		/// * [`Event::RecoveryCommitmentInvalidated`]
+		///
+		/// # Errors
+		/// * [`Error::NotAuthorizedRecoveryProvider`] - Caller is not an approved Recovery Provider.
+		/// * [`Error::ProviderNotRegistered`] - Recovery Provider is not registered.
+		/// * [`Error::NoRecoveryCommitment`] - No Recovery Commitment exists for the MSA.
+		/// * [`Error::InvalidRecoveryCommitment`] - Recovery Commitment does not match.
+		///
+		#[pallet::call_index(18)]
+		#[pallet::weight(T::WeightInfo::recover_account())]
+		pub fn recover_account(
+			origin: OriginFor<T>,
+			intermediary_hash_a: RecoveryHash,
+			intermediary_hash_b: RecoveryHash,
+			new_control_key_proof: MultiSignature,
+			add_key_payload: AddKeyData<T>,
+		) -> DispatchResult {
+			let provider_key = ensure_signed(origin)?;
+
+			// Get the provider's MSA ID and ensure they're a registered provider
+			let provider_msa_id = Self::ensure_valid_msa_key(&provider_key)?;
+			ensure!(
+				Self::is_registered_provider(provider_msa_id),
+				Error::<T>::ProviderNotRegistered
+			);
+
+			// Ensure the provider is approved for recovery operations
+			ensure!(
+				Self::is_approved_recovery_provider(&ProviderId(provider_msa_id)),
+				Error::<T>::NotAuthorizedRecoveryProvider
+			);
+
+			// Compute the recovery commitment from the intermediary hashes
+			let recovery_commitment: RecoveryCommitment =
+				Self::compute_recovery_commitment(intermediary_hash_a, intermediary_hash_b);
+
+			// Get the stored recovery commitment for this MSA
+			let stored_commitment: RecoveryCommitment =
+				MsaIdToRecoveryCommitment::<T>::get(add_key_payload.msa_id)
+					.ok_or(Error::<T>::NoRecoveryCommitment)?;
+
+			// Verify the computed recovery commitment matches the stored one
+			ensure!(
+				recovery_commitment == stored_commitment,
+				Error::<T>::InvalidRecoveryCommitment
+			);
+
+			// Implement the signature verification for the new control key, which is usually done by
+			// add_public_key_to_msa, but in this case we do not have the MSA owner key.
+			ensure!(
+				Self::verify_signature(
+					&new_control_key_proof,
+					&add_key_payload.new_public_key,
+					&add_key_payload
+				),
+				Error::<T>::NewKeyOwnershipInvalidSignature
```

Comment:
Not sure if some of these checks can be moved to extension side that might balance the weight better, i recall 6-9 rw